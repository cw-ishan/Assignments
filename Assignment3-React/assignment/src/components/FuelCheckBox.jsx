import axios from 'axios';
import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import fetchProducts from '../redux/products/productActions';

const styles={
    margin:'4%',
    fontSize:'20px',
    fontWeight:'100',
    fill:'#484848'
}
const updateUrl = (e) => {
    const fuelType = e.target.value
    const currentUrl = new URL(window.location.href);
    const searchParams = new URLSearchParams(currentUrl.search);
    const currentValues = searchParams.get('fuel')
    let updatedValues;
    if (e.target.checked) {
        updatedValues = currentValues ? `${currentValues}+${fuelType}` : fuelType;
    }
    else {
        const fuelArray = currentValues.split('+');
        updatedValues = fuelArray.filter((value) => value !== fuelType).join('+');
    }
    if (updatedValues) currentUrl.searchParams.set('fuel', updatedValues);
    else currentUrl.searchParams.delete('fuel')
    window.history.pushState(null, null, currentUrl);
}

function FuelCheckBox({ type, value}) {
    const dispatch=useDispatch();
    const ref=useRef(null);
    const products=useSelector(state=>state.products);
    useEffect(()=>{
        const fuelQuery=new URLSearchParams(window.location.search).get('fuel')
        if(fuelQuery && fuelQuery.includes(value)) ref.current.checked=true
        else ref.current.checked=false
    })
    const handleClick = (e) => {
        updateUrl(e);
        const params=decodeURIComponent(window.location.search);
        dispatch(fetchProducts('https://stg.carwale.com/api/stocks'+params))
    }
    return (
        <label style={styles}>
            <input type='checkbox' onClick={handleClick} value={value} ref={ref}/>
            {type}
        </label>
    )
}

export default FuelCheckBox