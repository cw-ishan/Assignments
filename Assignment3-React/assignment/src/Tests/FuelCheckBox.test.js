import { fireEvent, render,screen } from "@testing-library/react"
import FuelCheckBox from "../components/FuelCheckBox"
import { useDispatch } from "react-redux"
import { Checkbox } from "@mui/material"

jest.mock('react-redux')
beforeEach(()=>{
     const dispatch_function=(url)=>{}
     useDispatch.mockReturnValue(dispatch_function)
})
describe('FuelCheckBox Testing',()=>{
    it('Inital Testing to check working using fireEvent',()=>{
        render(<FuelCheckBox/>)
        const checkbox=screen.queryByRole('checkbox')
        fireEvent.click(checkbox)
        expect(checkbox.checked).toBeTruthy();
    })

    // To check weather url is updating, when checked box is clicked 
    it('Url Update on Checking',()=>{
        render(<FuelCheckBox value={1}/>)
        const checkbox=screen.queryByRole('checkbox')
        fireEvent.click(checkbox)
        expect(decodeURIComponent(window.location.search)).toContain('1')
    })
})

it("fuel checkbox snapshot",()=>{
    const {component}=render(<Checkbox/>)
    expect(component).toMatchSnapshot()
})