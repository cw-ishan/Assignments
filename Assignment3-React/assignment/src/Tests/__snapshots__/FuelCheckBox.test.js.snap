// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fuel checkbox snapshot 1`] = `undefined`;
